File: runners/gen.py
"""Deterministic expansion of generated test inputs.

Large stress cases are stored in problem JSON as a compact spec, e.g.

    {"__gen__": "ints", "n": 100000, "lo": -1000, "hi": 1000, "seed": 7}

and expanded here at run time. The same algorithm is implemented in
runners/gen.js, so Python and JavaScript see byte-identical inputs, and
tools/build_problems.py uses this module to compute expected outputs.

Never change the PRNG or a kind's semantics without rebuilding problems/ —
the stored expected outputs are tied to the exact sequence produced here.
"""

MASK = 0xFFFFFFFF
LOWER = "abcdefghijklmnopqrstuvwxyz"


class Rng:
    """mulberry32. Every step is a 32-bit operation, so the JavaScript port in
    runners/gen.js (using Math.imul) produces an identical stream."""

    def __init__(self, seed):
        self.state = seed & MASK

    def next(self):
        self.state = (self.state + 0x6D2B79F5) & MASK
        t = self.state
        t = ((t ^ (t >> 15)) * (t | 1)) & MASK
        t = (t ^ ((t + (((t ^ (t >> 7)) * (t | 61)) & MASK)) & MASK)) & MASK
        return (t ^ (t >> 14)) & MASK

    def below(self, n):
        return self.next() % n if n > 0 else 0

    def between(self, lo, hi):
        return lo + self.below(hi - lo + 1)

    def pick(self, seq):
        return seq[self.below(len(seq))]


def _shuffle(rng, items):
    for i in range(len(items) - 1, 0, -1):
        j = rng.below(i + 1)
        items[i], items[j] = items[j], items[i]
    return items


def _ints(spec, rng):
    lo = spec.get("lo", 0)
    hi = spec.get("hi", 100)
    return [rng.between(lo, hi) for _ in range(spec["n"])]


def _sorted_ints(spec, rng):
    values = _ints(spec, rng)
    if spec.get("unique"):
        values = sorted(set(values))
        # Refill so the requested length survives de-duplication.
        step = max(1, spec.get("step", 1))
        while len(values) < spec["n"]:
            values.append((values[-1] if values else spec.get("lo", 0)) + rng.between(1, step))
    values.sort(reverse=bool(spec.get("desc")))
    return values


def _perm(spec, rng):
    base = spec.get("base", 0)
    return _shuffle(rng, [base + i for i in range(spec["n"])])


def _string(spec, rng):
    alphabet = spec.get("alphabet", LOWER)
    return "".join(rng.pick(alphabet) for _ in range(spec["n"]))


def _grid(spec, rng):
    rows, cols = spec["rows"], spec["cols"]
    alphabet = spec.get("alphabet")
    out = []
    for _ in range(rows):
        if alphabet is not None:
            row = "".join(rng.pick(alphabet) for _ in range(cols))
            out.append(row if spec.get("asString", True) else list(row))
        else:
            out.append([rng.between(spec.get("lo", 0), spec.get("hi", 1)) for _ in range(cols)])
    return out


def _pairs(spec, rng):
    lo = spec.get("lo", 0)
    hi = spec.get("hi", 100)
    return [[rng.between(lo, hi), rng.between(lo, hi)] for _ in range(spec["n"])]


def _repeat(spec, _rng):
    return [spec.get("value", 0)] * spec["n"]


def _arange(spec, _rng):
    start = spec.get("start", 0)
    step = spec.get("step", 1)
    return [start + i * step for i in range(spec["n"])]


def _words(spec, rng):
    alphabet = spec.get("alphabet", LOWER)
    lo = spec.get("minLen", 1)
    hi = spec.get("maxLen", 8)
    return ["".join(rng.pick(alphabet) for _ in range(rng.between(lo, hi))) for _ in range(spec["n"])]


def _edges(spec, rng):
    """n directed edges over `nodes` vertices; acyclic when spec['dag']."""
    nodes = spec["nodes"]
    out = []
    for _ in range(spec["n"]):
        a = rng.below(nodes)
        b = rng.below(nodes)
        if spec.get("dag") and a == b:
            continue
        if spec.get("dag") and a < b:
            a, b = b, a
        out.append([a, b])
    return out


def _concat(spec, _rng):
    parts = [expand(part) for part in spec["parts"]]
    if parts and all(isinstance(part, str) for part in parts):
        return "".join(parts)
    out = []
    for part in parts:
        out.extend(part if isinstance(part, list) else [part])
    return out


def _strrepeat(spec, _rng):
    return spec.get("value", "a") * spec["n"]


def _shuffle_spec(spec, rng):
    value = expand(spec["of"])
    if isinstance(value, str):
        return "".join(_shuffle(rng, list(value)))
    return _shuffle(rng, list(value))


def _rotate(spec, _rng):
    value = expand(spec["of"])
    by = spec.get("by", 0) % (len(value) or 1)
    return value[by:] + value[:by]


def _intervals(spec, rng):
    """n intervals [start, start + length], possibly overlapping."""
    lo = spec.get("lo", 0)
    hi = spec.get("hi", 1000)
    max_len = spec.get("maxLen", 10)
    out = []
    for _ in range(spec["n"]):
        start = rng.between(lo, hi)
        out.append([start, start + rng.below(max_len + 1)])
    return out


def _disjoint(spec, rng):
    """n sorted, non-overlapping intervals — the shape "insert interval" expects."""
    out = []
    cursor = spec.get("start", 0)
    for _ in range(spec["n"]):
        cursor += rng.between(1, spec.get("gap", 5))
        end = cursor + rng.below(spec.get("maxLen", 5) + 1)
        out.append([cursor, end])
        cursor = end
    return out


def _sorted_grid(spec, rng):
    """A matrix whose rows ascend and whose rows ascend against each other."""
    value = spec.get("start", 0)
    out = []
    for _ in range(spec["rows"]):
        row = []
        for _ in range(spec["cols"]):
            value += rng.between(1, spec.get("step", 5))
            row.append(value)
        out.append(row)
    return out


def _zipsum(spec, _rng):
    """Element-wise sum of two generated lists — start times plus durations, say."""
    parts = [expand(part) for part in spec["parts"]]
    return [sum(values) for values in zip(*parts)]


def _duplicate(spec, _rng):
    times = spec.get("times", 2)
    return [value for value in expand(spec["of"]) for _ in range(times)]


def _sort_of(spec, _rng):
    return sorted(expand(spec["of"]), reverse=bool(spec.get("desc")))


def _without(spec, _rng):
    return [value for value in expand(spec["of"]) if value != spec["value"]]


def _weighted_edges(spec, rng):
    """n weighted edges over `nodes` vertices; a spanning path first when connected."""
    lo = spec.get("lo", 0)
    hi = spec.get("hi", 10000)
    nodes = spec["nodes"]
    out = []
    if spec.get("connected"):
        for i in range(nodes - 1):
            out.append([i, i + 1, rng.between(lo, hi)])
    for _ in range(spec.get("n", 0)):
        out.append([rng.below(nodes), rng.below(nodes), rng.between(lo, hi)])
    return out


def _knows(spec, rng):
    """The "who knows whom" matrix of the celebrity problem."""
    n = spec["n"]
    density = spec.get("density", 50)
    celebrity = spec.get("celebrity")
    matrix = [[1 if i == j else (1 if rng.below(100) < density else 0) for j in range(n)]
              for i in range(n)]
    if celebrity is not None:
        for i in range(n):
            matrix[celebrity][i] = 1 if i == celebrity else 0
            matrix[i][celebrity] = 1
    return matrix


def _adjacency(spec, rng):
    """Adjacency list for `nodes` vertices and `n` random edges."""
    nodes = spec["nodes"]
    out = [[] for _ in range(nodes)]
    for _ in range(spec.get("n", 0)):
        a = rng.below(nodes)
        b = rng.below(nodes)
        out[a].append(b)
        if not spec.get("directed"):
            out[b].append(a)
    return out


def _oplog_arg(spec, rng):
    if "int" in spec:
        lo, hi = spec["int"]
        return rng.between(lo, hi)
    if "choice" in spec:
        return rng.pick(spec["choice"])
    if "word" in spec:
        word = spec["word"]
        alphabet = word.get("alphabet", LOWER)
        length = rng.between(word.get("minLen", 1), word.get("maxLen", 6))
        return "".join(rng.pick(alphabet) for _ in range(length))
    raise ValueError("unknown oplog argument: {}".format(spec))


def _oplog(spec, rng):
    """An operation log for a design problem.

    Emits `[className, op, op, ...]` or the matching argument lists, chosen by
    `part`, so a single 5000-call log costs a few hundred bytes of JSON.

    A method may declare `delta` (how it changes the container's size) and
    `needs` (the minimum size it may be called at); when the log would call a
    method on too small a container, the first growing method is used instead.
    """
    methods = spec["methods"]
    pool = []
    for index, method in enumerate(methods):
        pool.extend([index] * method.get("weight", 1))
    grower = next((i for i, m in enumerate(methods) if m.get("delta", 0) > 0), 0)

    ops = [spec["cls"]]
    args = [spec.get("ctor", [])]
    size = 0
    for _ in range(spec["n"]):
        index = pool[rng.below(len(pool))]
        if size < methods[index].get("needs", 0):
            index = grower
        method = methods[index]
        ops.append(method["name"])
        args.append([_oplog_arg(arg, rng) for arg in method.get("args", [])])
        size = max(0, size + method.get("delta", 0))
    return ops if spec.get("part", "ops") == "ops" else args


KINDS = {
    "oplog": _oplog,
    "adj": _adjacency,
    "zipsum": _zipsum,
    "duplicate": _duplicate,
    "sortof": _sort_of,
    "without": _without,
    "wedges": _weighted_edges,
    "knows": _knows,
    "intervals": _intervals,
    "disjoint": _disjoint,
    "sortedgrid": _sorted_grid,
    "concat": _concat,
    "strrepeat": _strrepeat,
    "shuffle": _shuffle_spec,
    "rotate": _rotate,
    "ints": _ints,
    "sorted": _sorted_ints,
    "perm": _perm,
    "string": _string,
    "grid": _grid,
    "pairs": _pairs,
    "repeat": _repeat,
    "arange": _arange,
    "words": _words,
    "edges": _edges,
}


def expand(value):
    """Recursively replaces every {"__gen__": ...} node with its materialised value."""
    if isinstance(value, dict):
        kind = value.get("__gen__")
        if kind is None:
            return {k: expand(v) for k, v in value.items()}
        if kind not in KINDS:
            raise ValueError("unknown generator kind: {}".format(kind))
        return KINDS[kind](value, Rng(value.get("seed", 1)))
    if isinstance(value, list):
        return [expand(v) for v in value]
    return value

File: runners/gen.js
/**
 * Deterministic expansion of generated test inputs — JavaScript port of
 * runners/gen.py. Both must produce identical output for the same spec;
 * expected outputs in problems/*.json are computed from the Python side.
 */
'use strict';

const MASK = 0xffffffff;
const LOWER = 'abcdefghijklmnopqrstuvwxyz';

class Rng {
  constructor(seed) {
    this.state = seed >>> 0;
  }

  next() {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0;
    t = (t ^ (t + Math.imul(t ^ (t >>> 7), t | 61))) >>> 0;
    return (t ^ (t >>> 14)) >>> 0;
  }

  below(n) {
    return n > 0 ? this.next() % n : 0;
  }

  between(lo, hi) {
    return lo + this.below(hi - lo + 1);
  }

  pick(seq) {
    return seq[this.below(seq.length)];
  }
}

function shuffle(rng, items) {
  for (let i = items.length - 1; i > 0; i--) {
    const j = rng.below(i + 1);
    const tmp = items[i];
    items[i] = items[j];
    items[j] = tmp;
  }
  return items;
}

function oplogArg(spec, rng) {
  if (spec.int) return rng.between(spec.int[0], spec.int[1]);
  if (spec.choice) return rng.pick(spec.choice);
  if (spec.word) {
    const alphabet = spec.word.alphabet === undefined ? LOWER : spec.word.alphabet;
    const length = rng.between(spec.word.minLen === undefined ? 1 : spec.word.minLen,
                               spec.word.maxLen === undefined ? 6 : spec.word.maxLen);
    let out = '';
    for (let i = 0; i < length; i++) out += rng.pick(alphabet);
    return out;
  }
  throw new Error('unknown oplog argument: ' + JSON.stringify(spec));
}

const KINDS = {
  oplog(spec, rng) {
    const methods = spec.methods;
    const pool = [];
    methods.forEach((method, index) => {
      const weight = method.weight === undefined ? 1 : method.weight;
      for (let i = 0; i < weight; i++) pool.push(index);
    });
    let grower = methods.findIndex((m) => (m.delta || 0) > 0);
    if (grower < 0) grower = 0;

    const ops = [spec.cls];
    const args = [spec.ctor === undefined ? [] : spec.ctor];
    let size = 0;
    for (let i = 0; i < spec.n; i++) {
      let index = pool[rng.below(pool.length)];
      if (size < (methods[index].needs || 0)) index = grower;
      const method = methods[index];
      ops.push(method.name);
      args.push((method.args || []).map((arg) => oplogArg(arg, rng)));
      size = Math.max(0, size + (method.delta || 0));
    }
    return (spec.part === undefined ? 'ops' : spec.part) === 'ops' ? ops : args;
  },

  adj(spec, rng) {
    const out = [];
    for (let i = 0; i < spec.nodes; i++) out.push([]);
    for (let i = 0; i < (spec.n === undefined ? 0 : spec.n); i++) {
      const a = rng.below(spec.nodes);
      const b = rng.below(spec.nodes);
      out[a].push(b);
      if (!spec.directed) out[b].push(a);
    }
    return out;
  },

  zipsum(spec) {
    const parts = spec.parts.map(expand);
    const length = Math.min(...parts.map((p) => p.length));
    const out = new Array(length);
    for (let i = 0; i < length; i++) {
      let total = 0;
      for (const part of parts) total += part[i];
      out[i] = total;
    }
    return out;
  },

  duplicate(spec) {
    const times = spec.times === undefined ? 2 : spec.times;
    const out = [];
    for (const value of expand(spec.of)) {
      for (let i = 0; i < times; i++) out.push(value);
    }
    return out;
  },

  sortof(spec) {
    const values = expand(spec.of).slice();
    values.sort((a, b) => (spec.desc ? b - a : a - b));
    return values;
  },

  without(spec) {
    return expand(spec.of).filter((value) => value !== spec.value);
  },

  wedges(spec, rng) {
    const lo = spec.lo === undefined ? 0 : spec.lo;
    const hi = spec.hi === undefined ? 10000 : spec.hi;
    const out = [];
    if (spec.connected) {
      for (let i = 0; i < spec.nodes - 1; i++) out.push([i, i + 1, rng.between(lo, hi)]);
    }
    for (let i = 0; i < (spec.n === undefined ? 0 : spec.n); i++) {
      out.push([rng.below(spec.nodes), rng.below(spec.nodes), rng.between(lo, hi)]);
    }
    return out;
  },

  knows(spec, rng) {
    const n = spec.n;
    const density = spec.density === undefined ? 50 : spec.density;
    const matrix = [];
    for (let i = 0; i < n; i++) {
      const row = new Array(n);
      for (let j = 0; j < n; j++) row[j] = i === j ? 1 : rng.below(100) < density ? 1 : 0;
      matrix.push(row);
    }
    if (spec.celebrity !== undefined && spec.celebrity !== null) {
      for (let i = 0; i < n; i++) {
        matrix[spec.celebrity][i] = i === spec.celebrity ? 1 : 0;
        matrix[i][spec.celebrity] = 1;
      }
    }
    return matrix;
  },

  intervals(spec, rng) {
    const lo = spec.lo === undefined ? 0 : spec.lo;
    const hi = spec.hi === undefined ? 1000 : spec.hi;
    const maxLen = spec.maxLen === undefined ? 10 : spec.maxLen;
    const out = [];
    for (let i = 0; i < spec.n; i++) {
      const start = rng.between(lo, hi);
      out.push([start, start + rng.below(maxLen + 1)]);
    }
    return out;
  },

  disjoint(spec, rng) {
    const out = [];
    let cursor = spec.start === undefined ? 0 : spec.start;
    for (let i = 0; i < spec.n; i++) {
      cursor += rng.between(1, spec.gap === undefined ? 5 : spec.gap);
      const end = cursor + rng.below((spec.maxLen === undefined ? 5 : spec.maxLen) + 1);
      out.push([cursor, end]);
      cursor = end;
    }
    return out;
  },

  sortedgrid(spec, rng) {
    let value = spec.start === undefined ? 0 : spec.start;
    const out = [];
    for (let r = 0; r < spec.rows; r++) {
      const row = [];
      for (let c = 0; c < spec.cols; c++) {
        value += rng.between(1, spec.step === undefined ? 5 : spec.step);
        row.push(value);
      }
      out.push(row);
    }
    return out;
  },

  concat(spec) {
    const parts = spec.parts.map(expand);
    if (parts.length && parts.every((p) => typeof p === 'string')) return parts.join('');
    const out = [];
    for (const part of parts) {
      if (Array.isArray(part)) out.push(...part);
      else out.push(part);
    }
    return out;
  },

  strrepeat(spec) {
    return (spec.value === undefined ? 'a' : spec.value).repeat(spec.n);
  },

  shuffle(spec, rng) {
    const value = expand(spec.of);
    if (typeof value === 'string') return shuffle(rng, value.split('')).join('');
    return shuffle(rng, value.slice());
  },

  rotate(spec) {
    const value = expand(spec.of);
    const by = value.length ? ((spec.by === undefined ? 0 : spec.by) % value.length) : 0;
    return value.slice(by).concat(value.slice(0, by));
  },
  ints(spec, rng) {
    const lo = spec.lo === undefined ? 0 : spec.lo;
    const hi = spec.hi === undefined ? 100 : spec.hi;
    const out = new Array(spec.n);
    for (let i = 0; i < spec.n; i++) out[i] = rng.between(lo, hi);
    return out;
  },

  sorted(spec, rng) {
    let values = KINDS.ints(spec, rng);
    if (spec.unique) {
      values = [...new Set(values)].sort((a, b) => a - b);
      const step = Math.max(1, spec.step === undefined ? 1 : spec.step);
      while (values.length < spec.n) {
        const last = values.length ? values[values.length - 1] : spec.lo === undefined ? 0 : spec.lo;
        values.push(last + rng.between(1, step));
      }
    }
    values.sort((a, b) => (spec.desc ? b - a : a - b));
    return values;
  },

  perm(spec, rng) {
    const base = spec.base === undefined ? 0 : spec.base;
    const items = new Array(spec.n);
    for (let i = 0; i < spec.n; i++) items[i] = base + i;
    return shuffle(rng, items);
  },

  string(spec, rng) {
    const alphabet = spec.alphabet === undefined ? LOWER : spec.alphabet;
    let out = '';
    for (let i = 0; i < spec.n; i++) out += rng.pick(alphabet);
    return out;
  },

  grid(spec, rng) {
    const out = [];
    for (let r = 0; r < spec.rows; r++) {
      if (spec.alphabet !== undefined && spec.alphabet !== null) {
        let row = '';
        for (let c = 0; c < spec.cols; c++) row += rng.pick(spec.alphabet);
        out.push(spec.asString === false ? row.split('') : row);
      } else {
        const row = new Array(spec.cols);
        for (let c = 0; c < spec.cols; c++) {
          row[c] = rng.between(spec.lo === undefined ? 0 : spec.lo, spec.hi === undefined ? 1 : spec.hi);
        }
        out.push(row);
      }
    }
    return out;
  },

  pairs(spec, rng) {
    const lo = spec.lo === undefined ? 0 : spec.lo;
    const hi = spec.hi === undefined ? 100 : spec.hi;
    const out = new Array(spec.n);
    for (let i = 0; i < spec.n; i++) out[i] = [rng.between(lo, hi), rng.between(lo, hi)];
    return out;
  },

  repeat(spec) {
    return new Array(spec.n).fill(spec.value === undefined ? 0 : spec.value);
  },

  arange(spec) {
    const start = spec.start === undefined ? 0 : spec.start;
    const step = spec.step === undefined ? 1 : spec.step;
    const out = new Array(spec.n);
    for (let i = 0; i < spec.n; i++) out[i] = start + i * step;
    return out;
  },

  words(spec, rng) {
    const alphabet = spec.alphabet === undefined ? LOWER : spec.alphabet;
    const lo = spec.minLen === undefined ? 1 : spec.minLen;
    const hi = spec.maxLen === undefined ? 8 : spec.maxLen;
    const out = [];
    for (let i = 0; i < spec.n; i++) {
      const len = rng.between(lo, hi);
      let word = '';
      for (let c = 0; c < len; c++) word += rng.pick(alphabet);
      out.push(word);
    }
    return out;
  },

  edges(spec, rng) {
    const out = [];
    for (let i = 0; i < spec.n; i++) {
      let a = rng.below(spec.nodes);
      let b = rng.below(spec.nodes);
      if (spec.dag && a === b) continue;
      if (spec.dag && a < b) {
        const tmp = a;
        a = b;
        b = tmp;
      }
      out.push([a, b]);
    }
    return out;
  },
};

function expand(value) {
  if (Array.isArray(value)) return value.map(expand);
  if (value && typeof value === 'object') {
    const kind = value.__gen__;
    if (kind === undefined) {
      const out = {};
      for (const key of Object.keys(value)) out[key] = expand(value[key]);
      return out;
    }
    if (!KINDS[kind]) throw new Error('unknown generator kind: ' + kind);
    return KINDS[kind](value, new Rng(value.seed === undefined ? 1 : value.seed));
  }
  return value;
}

module.exports = { expand, Rng };
